File: Python/textbook_track/chapter03/ba3h.py
# Reconstruct a String from its k-mer Composition
import sys

from collections import defaultdict


def build_de_bruijn_graph(k_mers):
    d = defaultdict(list)
    for k_mer in k_mers:
        prefix = k_mer[:-1]
        suffix = k_mer[1:]
        d[prefix].append(suffix)
    return dict(d)


def calc_node_degrees(adjacency_list):
    degrees = defaultdict(int)
    for node, neighbours in adjacency_list.items():
        for neighbour in neighbours:
            degrees[node] += 1
            degrees[neighbour] -= 1
    return dict(degrees)


def find_eulerian_path_endpoints(adjacency_list):
    degrees = calc_node_degrees(adjacency_list)

    def find_node(value):
        for node, degree in degrees.items():
            if degree == value:
                return node

    path_start = find_node(value=1)
    path_end = find_node(value=-1)
    return path_start, path_end


def add_extra_edge(adjacency_list, path_start, path_end):
    neighbours = adjacency_list.setdefault(path_end, [])
    neighbours.append(path_start)


def find_node_with_unused_edges_on_cycle(adjacency_list, cycle):
    for ix, node in enumerate(cycle):
        if node in adjacency_list:
            return ix, node


def shift_cycle(cycle, ix):
    next_start_node = cycle[ix]
    _ = cycle.pop(-1)
    new_cycle = cycle[ix:] + cycle[:ix]
    new_cycle.append(next_start_node)
    return new_cycle


def extend_cycle(adjacency_list, start_node, cycle=None):
    if cycle is None:
        cycle = [start_node]

    current_node = start_node
    while current_node in adjacency_list:
        neighbours = adjacency_list[current_node]
        next_node = neighbours.pop(-1)
        if not neighbours:
            _ = adjacency_list.pop(current_node)

        cycle.append(next_node)
        current_node = next_node
    return cycle


def remove_extra_edge_from_cycle(cycle, path_start, path_end):
    for ix, (a, b) in enumerate(zip(cycle, cycle[1:])):
        if a == path_end and b == path_start:
            return cycle[ix+1:] + cycle[1:ix+1]
    return cycle


def find_eulerian_path(adjacency_list):
    path_start, path_end = find_eulerian_path_endpoints(adjacency_list)
    add_extra_edge(adjacency_list, path_start, path_end)

    start_node = min(adjacency_list)
    cycle = extend_cycle(adjacency_list, start_node)
    while adjacency_list:
        ix, new_start_node = find_node_with_unused_edges_on_cycle(adjacency_list, cycle)
        shifted_cycle = shift_cycle(cycle, ix)
        cycle = extend_cycle(adjacency_list, new_start_node, shifted_cycle)
    return remove_extra_edge_from_cycle(cycle, path_start, path_end)


def calc_string_spelled_by_a_genome_path(k_mers):
    text = list(k_mers[0])
    for k_mer in k_mers[1:]:
        text.append(k_mer[-1])
    return ''.join(text)


def main():
    data = sys.stdin.read().splitlines()
    k_mers = data[1:]
    adjacency_list = build_de_bruijn_graph(k_mers)
    path = find_eulerian_path(adjacency_list)
    result = calc_string_spelled_by_a_genome_path(path)
    print(result)


if __name__ == '__main__':
    main()

File: Scala/src/main/scala/TextbookTrack/Chapter03/BA3H.scala
package TextbookTrack.Chapter03

object BA3H {
  import scala.annotation.tailrec

  class DeBruijnGraph[T](adjacencyList: Map[T, List[T]])(implicit ord: Ordering[T]) {
    import DeBruijnGraph._

    def findEulerianPath(): List[T] = {
      @tailrec
      def loop(edges: Map[T, List[T]], cycle: List[T]): List[T] = {
        if (edges.isEmpty) cycle.reverse
        else {
          val (nextStartNode, ix): (T, Int) = findNodeWithUnusedEdges(edges, cycle)
          val shiftedCycle: List[T] = shiftCycle(cycle, ix, nextStartNode)
          val (remainingEdges, extendedCycle): (Map[T, List[T]], List[T]) =
            extendCycle(edges, nextStartNode, shiftedCycle)
          loop(remainingEdges, extendedCycle)
        }
      }

      findEulerianPathEndpoints(adjacencyList) match {
        case None => throw new Exception("The graph does not have an Eulerian path.")
        case Some((pathStart, pathEnd)) =>
          val extendedEdges: Map[T, List[T]] = addExtraEdge(adjacencyList, pathStart, pathEnd)
          val startNode: T = extendedEdges.keys.min
          val cycle: List[T] = loop(extendedEdges, List(startNode))
          removeExtraEdgeFromCycle(cycle, pathStart, pathEnd)
      }
    }
  }

  object DeBruijnGraph {
    def apply(kMers: List[String]): DeBruijnGraph[String] = {
      val adjacencyList: Map[String, List[String]] =
        kMers
          .map(kMer => (kMer.dropRight(1), kMer.drop(1)))
          .groupBy{ case (prefix, _) => prefix }
          .view
          .mapValues(_.map{ case (_, suffix) => suffix })
          .toMap
      new DeBruijnGraph[String](adjacencyList)
    }

    private def findNodeWithUnusedEdges[T](adjacencyList: Map[T, List[T]], cycle: List[T]): (T, Int) =
      cycle.iterator.zipWithIndex.find { case (node, _) => adjacencyList.contains(node) } match {
        case None => throw new Exception("The cycle must contain nodes with unused outgoing edges.")
        case Some((node, ix)) => (node, ix)
      }

    private def shiftCycle[T](cycle: List[T], ix: Int, startNode: T): List[T] = {
      val (first, second): (List[T], List[T]) = cycle.splitAt(ix + 1)
      startNode :: second ::: first.tail
    }

    private def extendCycle[T](adjacencyList: Map[T, List[T]],
                               startNode: T,
                               initialCycle: List[T]): (Map[T, List[T]], List[T]) = {
      @tailrec
      def loop(edges: Map[T, List[T]], cycle: List[T], currentNode: T): (Map[T, List[T]], List[T]) = {
        if (!edges.contains(currentNode)) (edges, cycle)
        else {
          edges.getOrElse(currentNode, Nil) match {
            case Nil => throw new Exception(s"The node $currentNode should have been removed earlier.")
            case nextNode :: rest =>
              val updatedEdges: Map[T, List[T]] =
                if (rest.isEmpty) edges.removed(currentNode) else edges.updated(currentNode, rest)
              loop(updatedEdges, nextNode :: cycle, nextNode)
          }
        }
      }

      loop(adjacencyList, initialCycle, startNode)
    }

    private def calcDegrees[T](adjacencyList: Map[T, List[T]]): Map[T, Int] =
      adjacencyList
        .iterator
        .flatMap{ case (node, neighbours) => neighbours.flatMap(neighbour => List((node, 1), (neighbour, -1)))}
        .toList
        .groupBy{ case (node, _) => node }
        .view
        .mapValues(_.map{ case (_, degree) => degree }.sum)
        .toMap

    private def findEulerianPathEndpoints[T](adjacencyList: Map[T, List[T]]): Option[(T, T)] = {
      val degrees: Map[T, Int] = calcDegrees(adjacencyList)
      def findNodeWithGivenDegree(value: Int): Option[T] =
        degrees.find{ case (_, degree) => degree == value }.map{ case (node, _) => node }

      val pathStart: Option[T] = findNodeWithGivenDegree(1)
      val pathEnd: Option[T] = findNodeWithGivenDegree(-1)
      for {
        start <- pathStart
        end <- pathEnd
      } yield (start, end)
    }

    private def addExtraEdge[T](adjacencyList: Map[T, List[T]], pathStart: T, pathEnd: T): Map[T, List[T]] = {
      val neighbours: List[T] = adjacencyList.getOrElse(pathEnd, Nil)
      adjacencyList.updated(pathEnd, pathStart :: neighbours)
    }

    private def removeExtraEdgeFromCycle[T](cycle: List[T], pathStart: T, pathEnd: T): List[T] = {
      val indexOfExtraEdge: List[Int] = (for {
        (List(a, b), ix) <- cycle.sliding(2).zipWithIndex
        if a == pathEnd && b == pathStart
      } yield ix).toList

      indexOfExtraEdge match {
        case ix :: _ => cycle.drop(ix + 1) ::: cycle.slice(1, ix + 1)
        case _ => throw new Exception("Extra edge appeared more than once or not at all.")
      }
    }
  }

  def calcStringSpelledByAGenomePath(kMers: List[String]): String = kMers match {
    case Nil => ""
    case x :: xs => (x.toList ::: xs.map(_.last)).mkString
  }

  def main(args: Array[String]): Unit = {
    val reader: Iterator[String] = scala.io.Source.stdin.getLines()
    val _: Int = reader.next().toInt
    val kMers: List[String] = reader.toList
    val graph = DeBruijnGraph(kMers)
    val path: List[String] = graph.findEulerianPath()
    val result: String = calcStringSpelledByAGenomePath(path)
    println(result)
  }
}
